Include CPU active time, lock wait time, and query status in ListQueries

thobe opened the pull request:
Sample output from `CALL dbms.listQueries`:

```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| queryId   | username | query                         | parameters | startTime                       | elapsedTime    | connectionDetails   | cpuTimeMillis | status                                                                                 | waitTimeMillis | metaData |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| "query-1" | ""       | "MATCH (n) SET n.v = n.v + 1" | {}         | "2017-01-03T16:38:29.213+01:00" | "00:00:01.507" | "embedded-session " | 856           | {waitTimeMillis -> 57, state -> "WAITING", resourceType -> "NODE", resourceIds -> [0]} | 57             | {}       |
| "query-4" | ""       | "CALL dbms.listQueries"       | {}         | "2017-01-03T16:38:30.719+01:00" | "00:00:00.001" | "embedded-session " | 1             | {state -> "RUNNING"}                                                                   | 0              | {}       |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

thobe committed:
Removing an element not in the list should not change the list

thobe committed:
Capture CPU time used by a query.

thobe committed:
Outline of tracing lock wait time into executing query status

thobe committed:
Implement same wait tracing behaviour in both lock managers

thobe committed:
Code cleanup

chrisvest commented:
What is it a sample output of, and can it include units?

praveenag commented:
@chrisvest This is the sample output of `dbms.listQueries` procedure. The units are in milliseconds. We chose not to explicitly call out units in the output to gather feedback.

praveenag committed:
fixed licence header

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -20,41 +20,64 @@
+    private final long cpuTimeNanosWhenQueryStarted;
     private Map<String,Object> metaData;
+    private volatile ExecutingQueryStatus status = ExecutingQueryStatus.RUNNING;
+    private volatile long waitTimeNanos;
```
It's a small nit, but I would prefer if the new instance fields get assigned in the constructor, so that's all kept together.

thobe replied:
for volatile fields that need a default value, I actually prefer it in the field declaration if possible, since it signals the non-null default nature of the field more clearly.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -111,17 +134,54 @@ public long startTime()
+        {
+            WAIT_TIME.addAndGet( ExecutingQuery.this, clock.nanos() - startTimeNanos );
+            status = previous;
+        };
```
I'm slightly irked that we allocate for the new status, and the lambda. The allocation for the status might be difficult to avoid, but it ought to be possible to avoid allocating the lambda, I think.

thobe replied:
Do you have a concrete suggestion for how? Making the status object also implement `Locks.WaitEvent` feels like conflating concerns, but that is the only way I can think of reducing the number of allocations here.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQueryStatus.java:
```diff
@@ -0,0 +1,82 @@
+import org.neo4j.storageengine.api.lock.ResourceType;
+import org.neo4j.time.SystemNanoClock;
+
+abstract class ExecutingQueryStatus
```
Please add a few lines of javadocs to this, describing what it's used for.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQueryStatus.java:
```diff
@@ -0,0 +1,82 @@
+        @Override
+        Map<String,Object> toMap( SystemNanoClock clock )
+        {
+            return Collections.singletonMap( "state", "RUNNING" );
```
I'm not sure how often this is called, but in principle it could be a constant as well.

thobe replied:
It is called when the status is inspected, so not on the hot path, but yes, it could be made constant.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/ExecutingQueryList.java:
```diff
@@ -19,21 +19,51 @@
+    };
 
-    class Entry implements ExecutingQueryList
+    abstract <T> T reduce( T defaultValue, Function<ExecutingQuery,T> accessor, BiFunction<T,T,T> combinator );
```
The `reduce` method could be moved above the `EMPTY` declaration.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/CombinedTracer.java:
```diff
@@ -0,0 +1,89 @@
+
+import org.neo4j.storageengine.api.lock.ResourceType;
+
+final class CombinedTracer implements Locks.Tracer
```
I would like to see some javadocs on this class as well.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/CombinedTracer.java:
```diff
@@ -0,0 +1,89 @@
+        {
+            tracers = Arrays.copyOf( this.tracers, this.tracers.length + 1 );
+            tracers[this.tracers.length] = tracer;
+        }
```
How big of an `n` do we expect to see here?

thobe replied:
if by `n` you mean the length of the `tracers` array, I'd anticipate the size to be(come) `3` if this code ever executes, typically I'd expect the size to be `2` in instances of `CombinedTracer`, do you think we should make it tree structured instead of flat? (since that would optimize for the expected size of 2)

chrisvest replied:
For such small N I don't have a strong opinion. An inlined tree would be one less dependent load. I'm not sure there's enough (or any) measurable difference in performance to justify this, but it may be a cleaner design. I'm fine either way.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -126,4 +126,43 @@ void visit( ResourceType resourceType, long resourceId, String description, long
 
     void close();
+
+    interface Tracer
```
I would prefer to call it `LockTracer`. I don't like how the monitors system has caused a proliferation of inner interfaces all called `Monitor`.

chrisvest replied:
Can we also make it a top-level interface? That would make it shorter to reference in other places in the PR.

thobe replied:
I always think of the names of nested types as being prefixed with the name of the outer type, making the name here `Locks.Tracer`, but when I wrote it I fully anticipated it to be moved to the top level and renamed `LocksTracer` at some point. The main thing that stopped me is the package structure mess.

chrisvest commented on enterprise/kernel/src/test/java/org/neo4j/kernel/enterprise/builtinprocs/ListQueriesProcedureTest.java:
```diff
@@ -0,0 +1,158 @@
+        //        Object startTime = row1.get( "startTime" );
+        //        Object elapsedTime = row1.get( "elapsedTime" );
+        //        Object connectionDetails = row1.get( "connectionDetails" );
+        //        Object metaData = row1.get( "metaData" );
```
What is this commented out stuff?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/api/ExecutingQuery.java:
```diff
@@ -20,41 +20,64 @@
+    private static final AtomicLongFieldUpdater<ExecutingQuery> WAIT_TIME =
+            newUpdater( ExecutingQuery.class, "waitTimeNanos" );
     private final long queryId;
+    private final Locks.Tracer lockTracer = ExecutingQuery.this::waitForLock;
```
Tracer implementations are meant to be changed via plug-ins, and the concrete implementation decided by `Tracers`. Hard coding the implementation here goes against this. For this reason, tracers should also not be relied upon for correctness; which is another mistake we've made with monitors. I think you should allow `Tracers` to decide what implementation to use, and then you can move the clock state into the default implementation. I would not be against adding a method to the interface to query the accumulated wait time, as long as you can accept that an alternative implementation might not track it and instead always report zero.

thobe replied:
I do not fully agree.

I agree that it should be possible to have a tracer provided through configuration. But I don't think that excludes being able to use the event based tracer pattern for other purposes as well. The important part of that is just that if one is needed for some feature, it shouldn't disable other features. This is one reason why we included the `CombinedTracer` so that the configured system-tracer can be combined with the query specific one(s).

thobe replied:
The place where a system-configured `Locks.Tracer` should be injected is in `KernelStatement` in the `lockTracer()` method, instead of the hard-coded `Locks.Tracer.NONE` there.

chrisvest reviewed (changes requested):
The overall idea is good, but please take a look at my comments. There are a couple of things I would like to see changed.

chrisvest commented:
As for the units, we could either add them as part of the column names in the `listQueries` output, or we could add columns dedicated to the units. Both of these solutions would make it easier for people to see what units are being reported, without compromising the machine readability. Adding the unit at the end of the value would require that the values be returned as strings which, while looking very nice to the human eye, would make it harder for machines to interpret the output.

chrisvest commented on enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java:
```diff
@@ -273,30 +288,46 @@ public void acquireExclusive( ResourceType resourceType, long... resourceIds ) t
-                    // Given a grace period of tries (to try and not starve readers), grab an update lock and wait
-                    // for it to convert to an exclusive lock.
-                    if ( tries > 50 && existingLock instanceof SharedLock )
+                Locks.WaitEvent waitEvent = null;
```
There are two empty lines in succession here. They are causing the build to fail on a checkstyle error.

chrisvest commented:
@thobe Now also needs a rebase and updated license headers.

thobe committed:
Code cleanup and some javadocs

thobe committed:
Combine WaitingOnLock status with Locks.WaitEvent to save allocation

thobe committed:
Update license headers

thobe committed:
Add time units to the fields

thobe committed:
Rename Locks.Tracer to LockTracer

Moves it to its own file. Also moves Locks.WaitEvent to LockWaitEvent.

thobe committed:
Expose LockTracer through TracerFactory

This allows a TracerFactory to provide a custom LockTracer that will
trace all lock wait events in the entire system (not just for a specific
query). This "system level" LockTracer will combine with the query level
LockTracer(s) such that query level lock wait times is always collected.

thobe committed:
All time fileds should be in millis

thobe committed:
fix test failure
